session.go

```diff
@@ -766,40 +766,55 @@ func (s *session) processTransportParameters(params *handshake.TransportParamete
 
 func (s *session) sendPackets() error {
 	s.pacingDeadline = time.Time{}
-	if !s.sentPacketHandler.SendingAllowed() { // if congestion limited, at least try sending an ACK frame
-		return s.maybeSendAckOnlyPacket()
+
+	sendMode := s.sentPacketHandler.SendMode()
+	if sendMode == ackhandler.SendNone { // shortcut: return immediately if there's nothing to send
+		return nil
 	}
+
 	numPackets := s.sentPacketHandler.ShouldSendNumPackets()
 	var numPacketsSent int
-	// Send retransmissions, until
-	// * we're congestion limited, or
-	// * there are no more retransmissions, or
-	// * the maximum number of packets was reached
-	for ; numPacketsSent < numPackets; numPacketsSent++ {
-		sentPacket, err := s.maybeSendRetransmission()
-		if err != nil {
-			return err
+sendLoop:
+	for {
+		switch sendMode {
+		case ackhandler.SendNone:
+			break sendLoop
+		case ackhandler.SendAck:
+			// We can at most send a single ACK only packet.
+			// There will only be a new ACK after receiving new packets.
+			return s.maybeSendAckOnlyPacket()
```

Comment:
So just to be clear, return here means that ACK-only packets don't contribute to pacing.